### PR TITLE
Fixing guidelines for the scheduler. 

### PR DESCRIPTION
**Pull request by indualagarsamy**
Relates to: https://github.com/Particular/Requirements/issues/188
Clarified the limitations of the scheduler. 

**Questions:**
- Do you think its better to separate out the last section to its own article?

### CODE REVIEW

**Comment by indualagarsamy** [comment]
@Particular/documentation - can you review what i currently have?

File: nservicebus/scheduling-with-nservicebus.md [hunk 1]
```diff
@@ -5,19 +5,19 @@ tags:
...
+- As soon as your task starts to get some branching logic (`if` or `switch` statements) or you start to add business logic, instead of a simple Bus.Send() or a Bus.SendLocal(), you should consider moving to a [saga](sagas-in-nservicebus.md) .
 
-Do not add any business logic in your tasks; instead, use [sagas](sagas-in-nservicebus.md).
+- Often times, rather than poll for a certain state using the Scheduler API, you can use an eventing model instead, where an event is published when that state transition occurs and take the necessary action in that event message handler.
 
-### Push versus Poll
+## How the scheduler works
 
-In the scenario where you are "polling" for some change in information state you should instead consider a "push" approach.
+The scheduler holds a all list of tasks scheduled in a non-durable in-memory dictionary. In version 3 and 4 tasks are scoped per `AppDomain`. In version 5 they are scoped per Bus instance.
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: "holds a all list" should be "holds a list"

File: nservicebus/scheduling-with-nservicebus.md [hunk 2]
```diff
@@ -5,19 +5,19 @@ tags:
...
 
-The scheduling API is often used to poll state, for example the file system or an external webservice. In these scenarios it is much more efficient to adopt a push based model. In the case of the file system you should instead consider a [FileSystemWatcher](http://msdn.microsoft.com/en-us/library/system.io.filesystemwatcher.aspx). In the case a webservice should instead consider a [WebHook](http://en.wikipedia.org/wiki/Webhook).
+When a new scheduled task is created it is given an unique identifier and stored in the endpoint's in-memory dictionary. The identifier for the task is sent in a message to the Timeout Manager for the message to be deferred for the specified time interval. When the specified time elapses, the Timeouts dispatcher returns the message containing the identifier to the endpoint with the scheduled task identifier, the endpoint uses that identifier to fetch and invoke the task from its internal list of tasks and executes them.
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: is the Saxon genitive form correct? "endpoint's in-memory" shouldn't is simply be "endpoint in-memory"

File: nservicebus/scheduling-with-nservicebus.md [hunk 3]
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
...
-
-## Assumptions
-
-- Scheduled tasks are created at endpoint startup using for example a class the implements the `IWantToRunWhenBusStartsAndStops` interface;
-- If the process restarts, all scheduled tasks are recreated and given new identifiers. Tasks scheduled before the restart will not be found and a message is written to the log. This is expected behavior;
-- A scheduled task will run forever, as long as the endpoint is running, there is no way to cancel it except skipping its execution each time is invoked;
+- Since the scheduler is non-durable, if the process restarts, all scheduled tasks (that are created during the endpoint's startup) are recreated and given new identifiers. Tasks scheduled before the restart that arrive at the endpoint queue will not be found and a message is written to the log as information.  
+- Already created schedule tasks cannot be canceled or changed.
+- Does not support a task to be started at a specified specific time.
 - Each scheduled task executes on a new thread using the `Task.Factory.StartNew(Action)` method, which means that there will be no transaction scope by default and it is up to you to create one if needed;
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: As far as I know we cannot assume that `Task.Factory.StartNew(Action)` will create a new thread, I'd rephrase it saying that each scheduled task will be executed in a new `Task` generated by the Task Parallel Library using `Task.Factory.StartNew(Action)`...

Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: @mauroservienti - will change it per your suggestion. I originally kept the statement as is (from the original doco).

File: nservicebus/scheduling-with-nservicebus.md (hunk 1 above)
Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: :) Will fix it.

File: nservicebus/scheduling-with-nservicebus.md (hunk 2 above)
Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: the in-memory dictionary belongs to the endpoint.

File: nservicebus/scheduling-with-nservicebus.md [hunk 4]
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
...
-- You will probably only do a `Bus.Send()` or `Bus.SendLocal()` in the task. The handler of the sent message will be invoked as all other handlers and if configured and supported by underlying transport will be wrapped in a transaction as usual;
+- Since the scheduler uses the queuing mechanism, it does have some side effects on the timeliness of scheduled tasks. When a task is scheduled to be run at a given time it is actual not "executed at that time", it is instead "queued at that time" to be executed. In most cases this distinction will have no noticeable effect on the behavior of the the scheduling API. However in high load systems the fact that a scheduled task is added to the back of the queue can result in a noticeable delay between the "time the task has been request to be run" and the "time the task is actually executed".
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: on the behavior of **the the** scheduling API

File: nservicebus/scheduling-with-nservicebus.md [hunk 5]
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
...
+- The scheduler API does not support scaling out the endpoint on the same machine or doing a side-by-side deployment of an endpoint. In both these cases, you have multiple endpoint instances sharing the same queue where the schedule task arrives.  As the scheduler is not-durable and the task details are stored in the endpoint instance that created the task, the task may or may not be found/executed if it is picked up the instance that did not originally create the schedule task. When a task specified by the ID cannot be found, it is no longer re-scheduled to be executed again.
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: I understood the above, but I'd say that it is not clear enough, the reader needs to connect the dots understanding that each time a new Task is scheduled a **new** task id is created, thus 2 instances of the same endpoint will create the same task with different ids, then the issue. Correct?

File: nservicebus/scheduling-with-nservicebus.md [hunk 6]
```diff
@@ -25,25 +25,20 @@ The difference between these examples is that in the latter a name is given for
...
 
-## When not to use it
+## Converting a schedule task into a saga
+To store these tasks durably in order to support either a scale-out scenario or a side-by-side deployment scenario, you can use the saga instead. You can think of the scheduler as a never ending saga.
```
Comment (mauroservienti) [inline nservicebus/scheduling-with-nservicebus.md]: use **a** saga instead

File: nservicebus/scheduling-with-nservicebus.md (hunk 5 above)
Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: I'll clarify it better. The issue is detailed in 188 (requirements)

File: Snippets/Snippets_4/Scheduling/Scheduling.cs [hunk 7]
```diff
@@ -14,7 +14,7 @@ public void ScheduleTask()
 
         // Name a schedule task and invoke it every 5 minutes
         Schedule.Every(TimeSpan.FromMinutes(5))
-            .Action("Task name", SomeCustomMethod);
+            .Action("MyCustomTask", SomeCustomMethod);
```
Comment (SeanFeldman) [inline Snippets/Snippets_4/Scheduling/Scheduling.cs]: Any particular reason for this rename? I'm just curious. Thank you.

Comment (indualagarsamy) [inline Snippets/Snippets_4/Scheduling/Scheduling.cs]: I wanted the saga name to correlate with the task name. In the last section, I'll make this more obvious. Thanks.

File: nservicebus/scheduling-with-nservicebus.md [hunk 8]
```diff
@@ -25,25 +25,26 @@ The difference between these examples is that in the latter a name is given for
...
-- A scheduled task will run forever, as long as the endpoint is running, there is no way to cancel it except skipping its execution each time is invoked;
-- Each scheduled task executes on a new thread using the `Task.Factory.StartNew(Action)` method, which means that there will be no transaction scope by default and it is up to you to create one if needed;
-- You will probably only do a `Bus.Send()` or `Bus.SendLocal()` in the task. The handler of the sent message will be invoked as all other handlers and if configured and supported by underlying transport will be wrapped in a transaction as usual;
-
-## When not to use it
+Using the same example usage of the scheduler API shown above, in order to convert it to a saga:
+1. Create a saga for the specified task name that is started by a message. Create a new message if necessary. In this example, the message that starts the saga is `StartMyCustomTaskSaga`. This message will be sent by the endpoint on startup, using a ` Bus.SendLocal(new StartMyCustomTaskSaga{TaskName = "StartupTask1"});`. If you did not use the API with the task name, name your saga appropriately. In this example, the saga is called `MyCustomTaskSaga`.
```
Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: Yes, will do.

File: nservicebus/scheduling-with-nservicebus.md [hunk 9]
```diff
@@ -25,25 +25,26 @@ The difference between these examples is that in the latter a name is given for
...
+- The scheduler API does not support scaling out the endpoint on the same machine or doing a side-by-side deployment of an endpoint. When you have multiple instances of the endpoint that's running on the same machine as is the case with these scenarios, these endpoint instances share the same input queue. Since each endpoint maintains its own created tasks in memory, when the specified time is up and the task is queued at the endpoint, any of the endpoint instance that is currently running can dequeue that message. If an endpoint that did not originally create this task happened to dequeue this message in order to be execute it, it will not find the task in its list. This will result in the task to not get executed and the task will also not be rescheduled again.
```
Comment (indualagarsamy) [inline nservicebus/scheduling-with-nservicebus.md]: Yeah, true. Will make it bold and a warning.

**Comment by indualagarsamy** [comment]
@mauroservienti @SeanFeldman - I have incorporated your comments. Also rebased with Master, so, should be able to merge this PR. Can you see if all your concerns are answered?
@Particular/documentation - Do you think we should split the saga example to a different article or leave it in this one? Let me know if this good, so I can get this merged soon.

**Comment by mauroservienti** [comment]
@indualagarsamy looks good.
@danielmarbach can you add more details on your suggestion?

**Comment by danielmarbach** [comment]
@mauroservienti which one?

**Comment by mauroservienti** [comment]
@danielmarbach sorry, this:

> Maybe also show how a saga which actually allow to cancel a scheduled task?

File: Snippets/Snippets_3/Scheduling/SchedulerSaga.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
+﻿namespace Snippets_3.Scheduling
+{
+    using System;
+    using NServiceBus;
+    using NServiceBus.Saga;
+
+    #region ScheduleTaskSaga
+
+    class MyCustomTaskSaga : Saga<MyCustomTaskSagaData>,
```
Comment (andreasohlund) [inline Snippets/Snippets_3/Scheduling/SchedulerSaga.cs]: Would there be better to make this a full sample instead?

**Comment by danielmarbach** [comment]
A cancel message which is sent to the saga and the saga is completed.

On Mon, Feb 23, 2015 at 2:44 PM, Mauro Servienti <notifications@github.com>
wrote:

> @danielmarbach <https://github.com/danielmarbach> sorry, this:
>
> Maybe also show how a saga which actually allow to cancel a scheduled task?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/571#issuecomment-75542476>
> .
>



-- 

Daniel Marbach

Solution Architect, Particular Software

Email:  daniel.marbach@particular.net

Web:    www.particular.net
<http://t.signauxneuf.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XX4RzDjlVS9QB88pTc_dW3Ljg7x56dWyBdgBXs402?t=http%3A%2F%2Fwww.particular.net%2F&si=4896787407044608&pi=33cc6202-1f31-4040-8d6c-e70641f44bba>

**Comment by mauroservienti** [comment]
@danielmarbach  as @andreasohlund is pointing out what about setting up a full sample?

**Comment by danielmarbach** [comment]
As an addition. Yeah great idea. Do you have time? I could do it tomorrow
or after tomorrow. Still have a few things to run today.

**Comment by mauroservienti** [comment]
not today, I'm trying to catch up with everything I lost in the last 4 days.

On Mon, Feb 23, 2015 at 2:53 PM, Daniel Marbach <notifications@github.com>
wrote:

> As an addition. Yeah great idea. Do you have time? I could do it tomorrow
> or after tomorrow. Still have a few things to run today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/571#issuecomment-75543792>
> .
>



-- 
Mauro Servienti | Solution Architect | Particular Software

**Comment by indualagarsamy** [comment]
@danielmarbach - I've updated the documentation to include the cancel task as a note to the Saga and moved around the `When Not to Use` to be more closer to the `Current Limitations` and also added another point in the When Not to use section regarding feature limitations.
